# Exploring eBay Car Sales Data
The aim of this project is to clean the data and analyze the included used car listings.

## Dataset 
In this project, I'll be working with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.

The dataset was originally scraped and uploaded to Kaggle by user orgesleka. This particular version of datasource was modified and provided as part of the DataQuest Data science course.

The data dictionary provided with data is as follows:
 - dateCrawled - When this ad was first crawled. All field-values are taken from this date.
 - name - Name of the car.
 - seller - Whether the seller is private or a dealer.
 - offerType - The type of listing
 - price - The price on the ad to sell the car.
 - abtest - Whether the listing is included in an A/B test.
 - vehicleType - The vehicle Type.
 - yearOfRegistration - The year in which the car was first registered.
 - gearbox - The transmission type.
 - powerPS - The power of the car in PS.
 - model - The car model name.
 - kilometer - How many kilometers the car has driven.
 - monthOfRegistration - The month in which the car was first registered.
 - fuelType - What type of fuel the car uses.
 - brand - The brand of the car.
 - notRepairedDamage - If the car has a damage which is not yet repaired.
 - dateCreated - The date on which the eBay listing was created.
 - nrOfPictures - The number of pictures in the ad.
 - postalCode - The postal code for the location of the vehicle.
 - lastSeenOnline - When the crawler saw this ad last online.

In [1]:
import pandas as pd
import numpy as np
pd.set_option('float_format', '{:f}'.format)

## Analysis
### Reading and overviewing the data

In [2]:
autos = pd.read_csv('autos.csv', encoding='Latin-1')
autos.info()
autos.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   

     gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0    manuell      158  andere  150,000km                    3      lpg   
1  automatik      286     7er  150,000km                    6   benzin   
2    manuell      102    golf   70,000km                    7   benzin   
3  automatik       71  fortwo   70,000km                    6   benzin   
4    manuell        0   focus  150,000km                    7   benzin   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50

We can alredy make some insights by just reading the csv file into a data structure. First off, the __"name"__ column, has many differently formatted strings, a couple things that pop out are all strings use "\_" separator and every name so far starts with the manufacturer of the car. Information, we can use.

Out of the 20 columns in this datasource most contain string data and only 5 are missing data, and only one is missing a huge amount __"notRepairedDamage"__ column only has 40171 entries instead of expected 50000, but this might not be a problem, as it could be the case, where cars without damage have an empty entry in the column.

First we will start with cleaning up the column names to make the dataset easier to work with.

### Cleaning the data
#### Formatting the columns

In [3]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [4]:
clean_columns = ['date_crawled', 'name', 'seller', 'offer_type', 'price',
                 'ab_test','vehicle_type', 'registration_year', 'gearbox', 
                 'power_ps', 'model', 'odometer', 'registration_month', 
                 'fuel_type', 'brand', 'unrepaired_damage', 'ad_created', 
                 'nr_of_pictures', 'postal_code', 'last_seen']
autos.columns = clean_columns
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

Here we made a few changes:
 - Made all column names follow the snakecase convention
 - Adjusted the names to more closely reflect the columns, e.g. __"dateCreated"__ to __"ad_created"__

#### Data exploration to determine what needs cleaning
Initially we are going to look for:
 - Text columns where all or almost all values are the same. These often can be dropped and will reduce data complexity
 - Any numeric data stored as text, that needs converting

In [5]:
autos.describe(include="all")

date_crawled         name  seller offer_type  price ab_test  \
count                 50000        50000   50000      50000  50000   50000   
unique                48213        38754       2          2   2357       2   
top     2016-03-09 11:54:38  Ford_Fiesta  privat    Angebot     $0    test   
freq                      3           78   49999      49999   1421   25756   
mean                    NaN          NaN     NaN        NaN    NaN     NaN   
std                     NaN          NaN     NaN        NaN    NaN     NaN   
min                     NaN          NaN     NaN        NaN    NaN     NaN   
25%                     NaN          NaN     NaN        NaN    NaN     NaN   
50%                     NaN          NaN     NaN        NaN    NaN     NaN   
75%                     NaN          NaN     NaN        NaN    NaN     NaN   
max                     NaN          NaN     NaN        NaN    NaN     NaN   

       vehicle_type  registration_year  gearbox     power_ps  model  \
count         44905       50000.000000    47320 50000.000000  47242   
unique            8                nan        2          nan    245   
top       limousine                nan  manuell          nan   golf   
freq          12859                nan    36993          nan   4024   
mean            NaN        2005.073280      NaN   116.355920    NaN   
std             NaN         105.712813      NaN   209.216627    NaN   
min             NaN        1000.000000      NaN     0.000000    NaN   
25%             NaN        1999.000000      NaN    70.000000    NaN   
50%             NaN        2003.000000      NaN   105.000000    NaN   
75%             NaN        2008.000000      NaN   150.000000    NaN   
max             NaN        9999.000000      NaN 17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 nan         7          40                 2   
top     150,000km                 nan    benzin  volkswagen              nein   
freq        32424                 nan     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures  postal_code            last_seen  
count                 50000    50000.000000 50000.000000                50000  
unique                   76             nan          nan                39481  
top     2016-04-03 00:00:00             nan          nan  2016-04-07 06:17:27  
freq                   1946             nan          nan                    8  
mean                    NaN        0.000000 50813.627300                  NaN  
std                     NaN        0.000000 25779.747957                  NaN  
min                     NaN        0.000000  1067.000000                  NaN  
25%                     NaN        0.000000 30451.000000                  NaN  
50%                     NaN        0.000000 49577.000000                  NaN  
75%                     NaN        0.000000 71540.000000                  NaN  
max                     NaN        0.000000 99998.000000                  NaN

In [6]:
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price  ab_test vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

Here we can see a couple of statictics emerging:
 - __"seller"__ and __"offer\_type"__ columns contain mostly one type of entry, "privat" and "Angebot" respectively, thus not providing useful data for further analysis
 - __"nr_of_pictures"__ column appears to only contain 0 values, and needs further investigation
 - __"price"__ and __"odometer"__ appear to be numerical colmuns that currently are stored as strings.
 
We're going to start by converting __"price"__ and __"odometer"__ into numeric values, by removing any non-numeric values from the strings and then converting the strings to appropriate numeric types.

#### Removing seller and offer_type data

In [7]:
autos = autos.drop(["seller", "offer_type"], axis=1)

#### Investigating nr_of_pictures data

In [8]:
autos["nr_of_pictures"].unique()

array([0])

In [9]:
autos["nr_of_pictures"].value_counts()

0    50000
Name: nr_of_pictures, dtype: int64

We can tell that the __"nr\_of\_pictures"__ column contains only 0 values, data like this is not useful to us so we're going to be dropping this column as well.

In [10]:
autos = autos.drop(["nr_of_pictures"], axis=1)

#### Exploring and cleaning the price and odometer data

In [11]:
autos["price"].unique()

array(['$5,000', '$8,500', '$8,990', ..., '$385', '$22,200', '$16,995'],
      dtype=object)

In [12]:
autos["price"] = autos["price"].str.replace("$", "")
autos["price"] = autos["price"].str.replace(",", "").astype(int)

In [13]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 17 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        50000 non-null  object
 1   name                50000 non-null  object
 2   price               50000 non-null  int64 
 3   ab_test             50000 non-null  object
 4   vehicle_type        44905 non-null  object
 5   registration_year   50000 non-null  int64 
 6   gearbox             47320 non-null  object
 7   power_ps            50000 non-null  int64 
 8   model               47242 non-null  object
 9   odometer            50000 non-null  object
 10  registration_month  50000 non-null  int64 
 11  fuel_type           45518 non-null  object
 12  brand               50000 non-null  object
 13  unrepaired_damage   40171 non-null  object
 14  ad_created          50000 non-null  object
 15  postal_code         50000 non-null  int64 
 16  last_seen           50

In [14]:
autos["odometer"].unique()

array(['150,000km', '70,000km', '50,000km', '80,000km', '10,000km',
       '30,000km', '125,000km', '90,000km', '20,000km', '60,000km',
       '5,000km', '100,000km', '40,000km'], dtype=object)

In [15]:
autos["odometer"] = autos["odometer"].str.replace("km", "")
autos["odometer"] = autos["odometer"].str.replace(",", "").astype(int)

In [16]:
autos.rename(columns={"odometer" : "odometer_km"}, inplace=True)

In [17]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 17 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   date_crawled        50000 non-null  object
 1   name                50000 non-null  object
 2   price               50000 non-null  int64 
 3   ab_test             50000 non-null  object
 4   vehicle_type        44905 non-null  object
 5   registration_year   50000 non-null  int64 
 6   gearbox             47320 non-null  object
 7   power_ps            50000 non-null  int64 
 8   model               47242 non-null  object
 9   odometer_km         50000 non-null  int64 
 10  registration_month  50000 non-null  int64 
 11  fuel_type           45518 non-null  object
 12  brand               50000 non-null  object
 13  unrepaired_damage   40171 non-null  object
 14  ad_created          50000 non-null  object
 15  postal_code         50000 non-null  int64 
 16  last_seen           50

Here for both __"price"__ and __"odometer"__ we checked for unique values, to determine the pattern that the strings are constructed. Then removed any non-numerical values and converted the string to integer value.

For the __"odometer"__ column we also renamed it to __"odometer\_km"__ to reflect the units of measurement used for the integer values.

We'll continue exploring the __"price"__ and the __"odometer\_km"__ columns to look for any outlier data.

In [18]:
autos["price"].unique().shape

(2357,)

In [19]:
autos["price"].describe()

count      50000.000000
mean        9840.043760
std       481104.380500
min            0.000000
25%         1100.000000
50%         2950.000000
75%         7200.000000
max     99999999.000000
Name: price, dtype: float64

Here we can tell, that there are some outliers in the pricing data. Minimum value in this column appears to be 0, while the maximum value is almost 100 million, both seem unrealistic. 

In [20]:
autos["price"].value_counts().sort_index(ascending=True).head(15)

0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
Name: price, dtype: int64

In [21]:
autos["price"].value_counts().sort_index(ascending=False).head(15)

99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
Name: price, dtype: int64

Looking at the top 15 and bottom 15 prices we can tell that:
 - There are over 1400 cars listed at \\$0
 - Lowest car price starts at \\$1, however ebay is an auction website and starting bid can indeed be \\$1
 - Top price is almost a \\$100mln. which does not seem realistic
 - Top 10 prices are unreal numbers like \\$12345678, while at 11 we have \\$350000 and prices steadily decline after that
 
 Having made these insights we're removing the rows containing the top 10 prices, as well as rows containging the \\$0 price

In [22]:
autos = autos[autos["price"].between(1, 350000)]

In [23]:
autos["price"].value_counts().sort_index(ascending=False).head()

350000    1
345000    1
299000    1
295000    1
265000    1
Name: price, dtype: int64

In [24]:
autos["price"].value_counts().sort_index(ascending=True).head()

1    156
2      3
3      1
5      2
8      1
Name: price, dtype: int64

In [25]:
autos["odometer_km"].unique().shape

(13,)

In [26]:
autos["odometer_km"].describe()

count    48565.000000
mean    125770.101925
std      39788.636804
min       5000.000000
25%     125000.000000
50%     150000.000000
75%     150000.000000
max     150000.000000
Name: odometer_km, dtype: float64

In [27]:
autos["odometer_km"].value_counts().sort_index(ascending=True)

5000        836
10000       253
20000       762
30000       780
40000       815
50000      1012
60000      1155
70000      1217
80000      1415
90000      1734
100000     2115
125000     5057
150000    31414
Name: odometer_km, dtype: int64

Applying the same technique we've used anaysing price data, we can see that __"odometer\_km"__ does not contain outlier data. All data looks realistic.

#### Exploring and cleaning the date data

There are 5 columns that should represent date values. Some of these columns were created by the crawler, some came from the website itself. We can differentiate by referring to the data dictionary:
-  __"date_crawled"__: added by the crawler
-  __"last_seen"__: added by the crawler
-  __"ad_created"__: from the website
-  __"registration_month"__: from the website
-  __"registration_year"__: from the website

Right now, the __"date_crawled", "last_seen",__ and __"ad_created"__ columns are all identified as string values by pandas. So let's start investigating by looking at how the strings were constructed

In [28]:
autos[['date_crawled','ad_created','last_seen']][0:5]

date_crawled           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50

In [29]:
date_crawled = autos['date_crawled'].str[:10].value_counts(normalize=True, 
                                            dropna=False)
date_crawled.sort_index(ascending=True)

2016-03-05   0.025327
2016-03-06   0.014043
2016-03-07   0.036014
2016-03-08   0.033296
2016-03-09   0.033090
2016-03-10   0.032184
2016-03-11   0.032575
2016-03-12   0.036920
2016-03-13   0.015670
2016-03-14   0.036549
2016-03-15   0.034284
2016-03-16   0.029610
2016-03-17   0.031628
2016-03-18   0.012911
2016-03-19   0.034778
2016-03-20   0.037887
2016-03-21   0.037373
2016-03-22   0.032987
2016-03-23   0.032225
2016-03-24   0.029342
2016-03-25   0.031607
2016-03-26   0.032204
2016-03-27   0.031092
2016-03-28   0.034860
2016-03-29   0.034099
2016-03-30   0.033687
2016-03-31   0.031834
2016-04-01   0.033687
2016-04-02   0.035478
2016-04-03   0.038608
2016-04-04   0.036487
2016-04-05   0.013096
2016-04-06   0.003171
2016-04-07   0.001400
Name: date_crawled, dtype: float64

From the __"date\_crawled"__ value distribution, we can tell, that the web scraping took roughly a month, starting 2016-03-05 and ending 2016-04-07. Every day roughtly the same amount of ads were scraped with a couple of exceptions during the first couple of days and the last couple of days.

In [33]:
ad_created = autos['ad_created'].str[:10].value_counts(normalize=True, 
                                            dropna=False)
ad_created.sort_index(ascending=True)

2015-06-11   0.000021
2015-08-10   0.000021
2015-09-09   0.000021
2015-11-10   0.000021
2015-12-05   0.000021
               ...   
2016-04-03   0.038855
2016-04-04   0.036858
2016-04-05   0.011819
2016-04-06   0.003253
2016-04-07   0.001256
Name: ad_created, Length: 76, dtype: float64

A large variety of ads were scraped during the month period. The biggest chunk are ads created roughly during the scraping period with a couple of older ads, oldest being almost 9 months old. 

In [31]:
last_seen = autos['last_seen'].str[:10].value_counts(normalize=True, 
                                            dropna=False)
last_seen.sort_index(ascending=True)

2016-03-05   0.001071
2016-03-06   0.004324
2016-03-07   0.005395
2016-03-08   0.007413
2016-03-09   0.009595
2016-03-10   0.010666
2016-03-11   0.012375
2016-03-12   0.023783
2016-03-13   0.008895
2016-03-14   0.012602
2016-03-15   0.015876
2016-03-16   0.016452
2016-03-17   0.028086
2016-03-18   0.007351
2016-03-19   0.015834
2016-03-20   0.020653
2016-03-21   0.020632
2016-03-22   0.021373
2016-03-23   0.018532
2016-03-24   0.019767
2016-03-25   0.019211
2016-03-26   0.016802
2016-03-27   0.015649
2016-03-28   0.020859
2016-03-29   0.022341
2016-03-30   0.024771
2016-03-31   0.023783
2016-04-01   0.022794
2016-04-02   0.024915
2016-04-03   0.025203
2016-04-04   0.024483
2016-04-05   0.124761
2016-04-06   0.221806
2016-04-07   0.131947
Name: last_seen, dtype: float64

Dates here correspond with the scraping period. During the majority of the period we can see that the number of ads for each day is very small, which means it represents ads that were removed, let's assume, for selling the car. The last three days have a large chunk of ads, this represents more the scraping period ending, rather than mass removal of advertisements.

In [34]:
autos["registration_year"].describe()

count   48565.000000
mean     2004.755421
std        88.643887
min      1000.000000
25%      1999.000000
50%      2004.000000
75%      2008.000000
max      9999.000000
Name: registration_year, dtype: float64

Looking a the __"registration\_year"__ we can see, that most of the data is a valid year, however, min and max values show us that there are indeed outliers, with min year being year 1000, roughly 900 years before the invention of the car and the max value being year 9999, rougly 8000 years in the future. Let's explore these values further.

In [39]:
autos["registration_year"].value_counts().sort_index(ascending=True).head(15)

1000    1
1001    1
1111    1
1800    2
1910    5
1927    1
1929    1
1931    1
1934    2
1937    4
1938    1
1939    1
1941    2
1943    1
1948    1
Name: registration_year, dtype: int64

In [40]:
autos["registration_year"].value_counts().sort_index(ascending=False).head(15)

9999       3
9000       1
8888       1
6200       1
5911       1
5000       4
4800       1
4500       1
4100       1
2800       1
2019       2
2018     470
2017    1392
2016    1220
2015     392
Name: registration_year, dtype: int64

Looking at top and bottom 15 values we can see that:
 - 4 cars are dated to be registered before the invention of the car
 - Roughly 1800 cars are dated as have be registered after the data was collected

In [42]:
autos = autos[autos["registration_year"].between(1900,2016)]
autos["registration_year"].value_counts(normalize=True).head(15)

2000   0.067608
2005   0.062895
1999   0.062060
2004   0.057904
2003   0.057818
2006   0.057197
2001   0.056468
2002   0.053255
1998   0.050620
2007   0.048778
2008   0.047450
2009   0.044665
1997   0.041794
2011   0.034768
2010   0.034040
Name: registration_year, dtype: float64

After removing the outlier data we can see, that majority of the cars were registered during the last 20 years (having in mind that the data was collected in 2016).

#### Exploring pricing data of the top brands

In [49]:
autos["brand"].describe()

count          46681
unique            40
top       volkswagen
freq            9862
Name: brand, dtype: object

In [50]:
autos["brand"].unique()

array(['peugeot', 'bmw', 'volkswagen', 'smart', 'ford', 'chrysler',
       'seat', 'renault', 'mercedes_benz', 'audi', 'sonstige_autos',
       'opel', 'mazda', 'porsche', 'mini', 'toyota', 'dacia', 'nissan',
       'jeep', 'saab', 'volvo', 'mitsubishi', 'jaguar', 'fiat', 'skoda',
       'subaru', 'kia', 'citroen', 'chevrolet', 'hyundai', 'honda',
       'daewoo', 'suzuki', 'trabant', 'land_rover', 'alfa_romeo', 'lada',
       'rover', 'daihatsu', 'lancia'], dtype=object)

In [58]:
brand_count = autos["brand"].value_counts(normalize=True)
print(brand_count)

volkswagen       0.211264
bmw              0.110045
opel             0.107581
mercedes_benz    0.096463
audi             0.086566
ford             0.069900
renault          0.047150
peugeot          0.029841
fiat             0.025642
seat             0.018273
skoda            0.016409
nissan           0.015274
mazda            0.015188
smart            0.014160
citroen          0.014010
toyota           0.012703
hyundai          0.010025
sonstige_autos   0.009811
volvo            0.009147
mini             0.008762
mitsubishi       0.008226
honda            0.007840
kia              0.007069
alfa_romeo       0.006641
porsche          0.006127
suzuki           0.005934
chevrolet        0.005698
chrysler         0.003513
dacia            0.002635
daihatsu         0.002506
jeep             0.002271
subaru           0.002142
land_rover       0.002099
saab             0.001649
jaguar           0.001564
daewoo           0.001500
trabant          0.001392
rover            0.001328
lancia      

In [62]:
brand_count[brand_count > 0.05].sum()

0.6818191555450827

We can see that the vast majority of brands fall below 5\% of total values, while the top 6 brands make up a vast majority (around 70\%) of total values. Going forward we're going to be exploring brands that make up 5\%+ of total values.

In [63]:
brand_count_index = brand_count[brand_count > 0.05].index
print(brand_count_index)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [65]:
mean_price = {}
for brand in brand_count_index:
    brand_set = autos[autos["brand"] == brand]
    mean_price[brand] = brand_set["price"].mean()
mean_price

{'volkswagen': 5402.410261610221,
 'bmw': 8332.820517811953,
 'opel': 2975.2419354838707,
 'mercedes_benz': 8628.450366422385,
 'audi': 9336.687453600594,
 'ford': 3749.4695065890287}

Of the top 6 car brands we can see a couple of trends emerging:
 - Cars from BMW, Audi and Mercedes Benz are genereally more expensive, ranging over \\$8000
 - Cars from Opel and Ford are generally cheaper, with prices in ranges less than \\$4000
 - Cars from Volkswagen are the most popular, but don't fall in the ranges above. Volkswagen cars are generally priced somewhere in the middle between the more expensive BMW, Mercedes and Audi and the cheaper Ford and Opel which might explain their popularity.

#### Exploring mileage data of the top brands

Let's use the same techique used to analise pricing to take a look at mileage data. We'll compare mileage to see if there is any correlation between mileage and pricing.

In [67]:
mean_mileage = {}
for brand in brand_count_index:
    brand_set = autos[autos["brand"] == brand]
    mean_mileage[brand] = brand_set["odometer_km"].mean()
mean_mileage

{'volkswagen': 128707.15879132022,
 'bmw': 132572.51313996495,
 'opel': 129310.0358422939,
 'mercedes_benz': 130788.36331334666,
 'audi': 129157.38678544914,
 'ford': 124266.01287159056}

In [68]:
bmp = pd.Series(mean_price)
bmm = pd.Series(mean_mileage)
#Making both into Series objects

In [69]:
#Making the mean pricing series into a dataframe
brand_mean_price_mileage = pd.DataFrame(bmp, columns=["mean_price"])

In [70]:
brand_mean_price_mileage["mean_mileage"] = bmm

In [71]:
brand_mean_price_mileage

mean_price  mean_mileage
volkswagen    5402.410262 128707.158791
bmw           8332.820518 132572.513140
opel          2975.241935 129310.035842
mercedes_benz 8628.450366 130788.363313
audi          9336.687454 129157.386785
ford          3749.469507 124266.012872

There is less variety in the mileage for the top 6 brands then there was in price. Generally we can see a slight trend emerging, where mileage of the more expensive brands is slightly higher, than the cheaper counterparts.

Data cleaning next steps:
 - Identify categorical data that uses german words, translate them and map the values to their english counterparts
 - Convert the dates to be uniform numeric data, so "2016-03-21" becomes the integer 20160321.
 - See if there are particular keywords in the name column that you can extract as new columns
Analysis next steps:
Find the most common brand/model combinations
Split the odometer_km into groups, and use aggregation to see if average prices follows any patterns based on the mileage.
How much cheaper are cars with damage than their non-damaged counterparts?